### ISSUE TITLE
Anti-ping default value(s) discussion

### ISSUE THREAD

**Issue by kamillentee** [ddnet/ddnet#3721]
Some month ago I was suggesting to enable antiping by default for everyone, deen told me half of the players have weird bugs with antiping so thats not possible. After talking with many players I need to agree on that. 
However when having a high ping (other countries, bad peering) even those players who are strictly against antiping enable it, so enabeling it at a certain ping should not be a problem.
Secondly I often see new people who dont know about antiping struggle with things like wallhammer or hammerhit, while having a stable ping between 50 and 90. When I suggest to turn antiping on they usually play much better. 
So after some tests I think 50 seems like a good base to enable antiping, I personally would prefer 40 already but 60 seems still good enough to have an actual cause.

**Comment by def-** [comment]
These kind of changes are always risky since changing the default will change everyone's value who had the previous default set. Having a ping close to 50 will also be annoying since it will switch between antiping on/off.

**Comment by fokkonaut** [comment]
Please no, for example my ping on german servers is 48 and it jumps up and down a bit, so that would be very annoying for player with a similar connection.

**Comment by kamillentee** [comment]
Ok I see that's an issue for certain players. To solve this concern, antiping should not disable itself before the ping lowers by 10 (maybe less?). How much does your ping jump around fokkonaut?
Changing the default value for everyone comes with a risk of course, but I don't really see it here as first the majority will not realize the change due to good ping and for the rest it's a good change. Try playing around with a dummy and without antiping with 60 ping, nobody can play with that. 
Does cl_antiping_limit disable antiping for those who enabled cl_antiping when the ping is below the activation thresold? It should not do that.

**Comment by fokkonaut** [comment]
I think its generally a bad idea to make this a default. There should at most be a warning or a hint to enable antiping if the ping differs a lot from usual pings. (warning should not occur everytime, just once or twice, not sure). Making this a default is not good at all imho.

**Comment by heinrich5991** [comment]
I believe we should just make `cl_antiping` a default (only) for new players.

I think we even have a mechanism for this (in order to not change the default for old players), something called `cl_config_version`.

**Comment by heinrich5991** [comment]
> These kind of changes are always risky since changing the default will change everyone's value who had the previous default set. Having a ping close to 50 will also be annoying since it will switch between antiping on/off.

That suggests that this approach to changing defaults is the wrong one. See the above post for the (IMO) proper fix.

**Comment by def-** [comment]
> I believe we should just make cl_antiping a default (only) for new players.

But the game looks really horrible with antiping. I think many will be put off instantly when they see tees flying around randomly all the time.

**Comment by trml** [comment]
Regarding cl_antiping_limit, having a ping that jumps sounds like it could create problems in general with it. Perhaps it would make sense to have a version of it that only predicts the difference between the limit and your ping (when your ping exceeds it)?

**Comment by heinrich5991** [comment]
> But the game looks really horrible with antiping. I think many will be put off instantly when they see tees flying around randomly all the time.

Hm. Is that your experience with antiping? My experience with antiping is not that tees fly around, but rather that the hook and hammer works consistently and when I see it. That is true for the low-ping servers I usually play on.

What do you worry about? Players joining high-ping servers? I'd say those would be currently unplayable unless you already know that you have to manually predict where tees will be in 200ms. Players joining low-ping servers? I don't think you really see tees jumping around there.

I'm happy that we're now at least discussing the actual issue and not some technicality on how to change a default.

**Comment by fokkonaut** [comment]
I personally only play with antiping when my ping gets higher than like 100 or even higher or whatever, 50 is way too low, even 60.
AntiPing should not be a feature beginners should learn to play with, because once you really attach to it you can't play without it anymore. Even on servers with a ping of 20 or 30 antiping changes the feeling, There should be just a notice or a tip that antiping MIGHT help you, because for me, it usually doesnt, I see tees flying around like deen.

**Comment by def-** [comment]
Yeah, I'd be ok with this notice, or even showing a dialog that lets you choose whether to enable antiping when you join a server with > 80/X ping.

**Comment by heinrich5991** [comment]
> because once you really attach to it you can't play without it anymore.

What's the problem with this?

> Even on servers with a ping of 20 or 30 antiping changes the feeling, 

Yes, it makes it consistent. I did not have to change my play to adapt for server ping anymore. This is great!

**Comment by Smetanolub** [comment]
you should make it default by 0, so its always enabled. even when you play on lan with ping 1, antiping shows exact position of other player, without antiping everything is delayed and you need to adapt

**Comment by kaitlynia** [comment]
"Would you like to enable antiping?

If you play without it, player positions will feel out of sync but appear more smooth and continuous.

With antiping, your gameplay will feel much more in sync but show jumpy movements due to position prediction.

[off comparison gif] [on comparison gif]

[toggle]"

**Comment by AssassinTee** [comment]
What people do not get, is that you set the value to your [jitter](https://en.wikipedia.org/wiki/Jitter) and not your actual ping. For most people an AB value of 10 (ms) is totally enough. I personally play around with 4-8.

Since this is entirely depending on your connection, no matter what value we choose **this is going to be wrong for one or the other player**.

If you set the delay too high, it also becomes noticeable and I do notice an antiping value of 50 compared to 10.

Here a better issue: Calculate the antiping value while playing and adjust it automatically.

> That suggests that this approach to changing defaults is the wrong one. See the above post for the (IMO) proper fix

Yes

> Yeah, I'd be ok with this notice, or even showing a dialog that lets you choose whether to enable antiping when you join a server with > 80/X ping.

I was almost going to close this issue, but this would be the action I'd take now

**Comment by ddadikted** [comment]
Why not just leave it to a player preference? Turn off antiping and prediction margin by default, those who need it will enable it. I also dont get why we cant put prediction_margin to 0, 1 is minimum... i prefer less stability over performance, lower ping is better than stable and bigger(with margin)for me

**Comment by AssassinTee** [comment]
> Why not just leave it to a player preference? Turn off antiping and prediction margin by default, those who need it will enable it. I also dont get why we cant put prediction_margin to 0, 1 is minimum... i prefer less stability over performance, lower ping is better than stable and bigger(with margin)for me

Antiping is disabled by default, see #10644 and https://github.com/ddnet/ddnet/blob/beb46cf257254aa89226edc027e47f9804d948b4/src/engine/shared/config_variables.h#L19.

> also dont get why we cant put prediction_margin to 0, 1 is minimum...

Because 0 would equal the antiping disabled case

**Comment by KebsCS** [comment]
Antiping is good enough in 2025 to be enabled by default and almost everything is predicted

**Comment by sjrc6** [comment]
> Antiping is good enough in 2025 

This implies that it has improved, but it's more or less the exact same experience since it was added. Prediction is already enabled by default.

**Comment by AssassinTee** [comment]
> This implies that it has improved, but it's more or less the exact same experience since it was added. Prediction is already enabled by default.

Attention @sjrc6 , `prediction` is enabled by default, but `antping` is not, see https://github.com/ddnet/ddnet/blob/beb46cf257254aa89226edc027e47f9804d948b4/src/engine/shared/config_variables.h#L19

**Comment by AssassinTee** [comment]
# Antiping implementation overview:

### Weapons:

teleweapons: 🔴 #10260
jetpack gun: 🟢 
normal weapons: 🟢 
ninja: 🟢
grenade explosions: 🟡 (animation is off) #11360
freeze projectiles (shotgun pallets): 🟢
plasma bullets (laser head thingies):  🟢 #10249

### Map:

freeze-lasers: 🔴  
doors: 🟢
draggers: 🟢
bullets: 🟢
hearts: 🟢
shields: 🟢
stoppers: 🟢 
tele-hook 🔴 #9269 #11321
teleporters 🔴 #237 #11321
envelope animations: 🟢 #11299 
weapon spawning: ❓ 

Game mechanics:

death: 🔴 #11288 
double jump: 🟡 #10921
moving: 🟢
hooking: 🟢 (mostly)
jumping: 🟢 
dummy: 🟢 (with #11371 )

Other:

sounds (and events): 🟡#10280 #11360, implemented but still buggy
particles: 🟡 #10280 #11360, implemented but still buggy

> Antiping is good enough in 2025 to be enabled by default and almost everything is predicted

**Comment by KebsCS** [comment]
So looks like we're blocked by https://github.com/ddnet/ddnet/issues/10275 for most of the gameplay prediction
